src/fu-self-test.c

```diff
@@ -3756,6 +3756,106 @@ fu_device_list_counterpart_func(gconstpointer user_data)
 	g_assert_true(fu_device_has_counterpart_guid(device2, "runtime"));
 }
 
+static void
+fu_device_list_equivalent_id_func(gconstpointer user_data)
+{
+	FuTest *self = (FuTest *)user_data;
+	g_autoptr(FuDevice) device1 = fu_device_new(self->ctx);
+	g_autoptr(FuDevice) device2 = fu_device_new(self->ctx);
+	g_autoptr(FuDevice) device3 = fu_device_new(self->ctx);
+	g_autoptr(FuDevice) device_expected1 = NULL;
+	g_autoptr(FuDevice) device_expected2 = NULL;
+	g_autoptr(FuDevice) device_expected3 = NULL;
+	g_autoptr(FuDevice) device_expected4 = NULL;
+	g_autoptr(FuDeviceList) device_list = fu_device_list_new();
+	g_autoptr(GPtrArray) active1 = NULL;
+	g_autoptr(GPtrArray) active2 = NULL;
+	g_autoptr(GPtrArray) active3 = NULL;
+	g_autoptr(GPtrArray) active4 = NULL;
+	g_autoptr(GPtrArray) all1 = NULL;
+	g_autoptr(GPtrArray) all2 = NULL;
+	g_autoptr(GPtrArray) all3 = NULL;
+	g_autoptr(GPtrArray) all4 = NULL;
+	g_autoptr(GError) error = NULL;
+
+	/* simulate the device with 3 different physical connection points.
+	 * Using the equivalent ID to mark all connections belong to the same device.
+	 */
+	fu_device_set_id(device1, "8e9cb71aeca70d2faedb5b8aaa263f6175086b2e");
+	fu_device_set_equivalent_id(device1, "uniqueid");
```

Comment:
this makes me a bit confused; the equivalent_id is supposed to be the id of an existing device -- e.g. `fu_device_set_equivalent_id(device2, "8e9cb71aeca70d2faedb5b8aaa263f6175086b2e");` could make sense -- this random string less so.

If we're just tying devices together with a random shared string that's a counterpart GUID. Let me see if I can take your tests (which I think really clearly explain the problem, good job!) and use the counterpart logic to do the same thing.

Reply:
@hughsie please check this description first: https://github.com/fwupd/fwupd/issues/7769
Probably the equivalent ID is just a workaround and shouldn't be used, as I'm willing to use it